distributed/tests/test_active_memory_manager.py

```diff
@@ -385,6 +419,123 @@ async def test_drop_prefers_paused_workers(c, s, *workers):
     assert ws not in ts.who_has
 
 
+@pytest.mark.slow
+@gen_cluster(client=True, config=demo_config("drop"))
+async def test_drop_with_paused_workers_with_running_tasks_1(c, s, a, b):
+    """If there is exactly 1 worker that holds a replica of a task that isn't paused or
+    retiring, and there are 1+ paused/retiring workers with the same task, don't drop
+    anything.
+
+    Use case 1 (don't drop):
+    a is paused and with dependent tasks executing on it
+    b is running and has no dependent tasks
+    """
+    x = (await c.scatter({"x": 1}, broadcast=True))["x"]
+    y = c.submit(slowinc, x, delay=2, key="y", workers=[a.address])
+    while "y" not in a.tasks or a.tasks["y"].state != "executing":
+        await asyncio.sleep(0.01)
+    a.memory_pause_fraction = 1e-15
+    while s.workers[a.address].status != Status.paused:
+        await asyncio.sleep(0.01)
+    assert s.tasks["y"].state == "processing"
+    assert a.tasks["y"].state == "executing"
+
+    s.extensions["amm"].run_once()
+    await y
+    assert len(s.tasks["x"].who_has) == 2
+
+
+@gen_cluster(client=True, config=demo_config("drop"))
+async def test_drop_with_paused_workers_with_running_tasks_2(c, s, a, b):
+    """If there is exactly 1 worker that holds a replica of a task that isn't paused or
+    retiring, and there are 1+ paused/retiring workers with the same task, don't drop
+    anything.
+
+    Use case 2 (drop from a):
+    a is paused and has no dependent tasks
+    b is running and has no dependent tasks
+    """
+    x = (await c.scatter({"x": 1}, broadcast=True))["x"]
+    a.memory_pause_fraction = 1e-15
+    while s.workers[a.address].status != Status.paused:
+        await asyncio.sleep(0.01)
+
+    s.extensions["amm"].run_once()
+    await asyncio.sleep(0.2)
+    assert {ws.address for ws in s.tasks["x"].who_has} == {b.address}
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize("pause", [True, False])
+@gen_cluster(client=True, config=demo_config("drop"))
+async def test_drop_with_paused_workers_with_running_tasks_3_4(c, s, a, b, pause):
+    """If there is exactly 1 worker that holds a replica of a task that isn't paused or
+    retiring, and there are 1+ paused/retiring workers with the same task, don't drop
+    anything.
+
+    Use case 3 (drop from b):
+    a is paused and with dependent tasks executing on it
+    b is paused and has no dependent tasks
```

Comment:
Paused workers don't steal or start running tasks though. They exclusively finish their currently running (not queued) tasks. So B will never steal from A because B is paused.

> I guess it just feels odd to me to go all-in on a paused worker (both since it's clearly already under load, and it's more liable to crash). 

We're not adding stuff to A - we're just not deleting from it. Yes the tasks running on A could worsen the memory pressure but there's nothing saying that B hasn't got unrelated tasks running on it too.

Also note that I plan, in a future PR, for a worker to send back to the scheduler all of its queued tasks as soon as it enters paused state.

Not sure if this addresses your comment?

Reply:
> So B will never steal from A because B is paused

I didn't know this. I would have thought paused workers were the top candidates to steal from!

Because the current pause implementation is so problematic, I guess my gut reaction here is "a paused worker is soon to be a dead worker, so don't make it the only thing holding onto a key that we still need. Let's keep a backup of the key on a non-paused worker as an insurance policy until the paused worker proves able to un-pause itself. Then we can reconsider dropping the replica."

Not that I think this needs to happen in this PR, just for my own understanding.

> Also note that I plan, in a future PR, for a worker to send back to the scheduler all of its queued tasks as soon as it enters paused state.

This would help a lot, since we'll be able to make a much more informed decision.

Reply:
> I didn't know this. I would have thought paused workers were the top candidates to steal from!

I think you got it the other way around. The stealing doesn't happen because the *recipient* of the stolen tasks is paused.
There is currently no logic to steal *more* tasks from paused workers - as a matter of fact, until very recently the scheduler didn't even know that a worker was paused.